File: src/main/java/com/soapdogg/lispInterpreter/singleton/PrinterSingleton.java
package com.soapdogg.lispInterpreter.singleton;

import com.soapdogg.lispInterpreter.printer.*;
import lombok.Getter;

@Getter
public enum PrinterSingleton {
    INSTANCE;

    private final AtomNodePrinter atomNodePrinter;
    private final DotNotationExpressionNodePrinter dotNotationExpressionNodePrinter;
    private final DotNotationPrinter dotNotationPrinter;
    private final ListNotationExpressionNodePrinter listNotationExpressionNodePrinter;
    private final ListNotationPrinter listNotationPrinter;

    PrinterSingleton() {
        atomNodePrinter = new AtomNodePrinter();
        dotNotationExpressionNodePrinter = new DotNotationExpressionNodePrinter(
            DeterminerSingleton.INSTANCE.getExpressionNodeDeterminer(),
            atomNodePrinter
        );
        dotNotationPrinter = new DotNotationPrinter(
            DeterminerSingleton.INSTANCE.getExpressionNodeDeterminer(),
            atomNodePrinter,
            dotNotationExpressionNodePrinter
        );
        listNotationExpressionNodePrinter = new ListNotationExpressionNodePrinter(
            DeterminerSingleton.INSTANCE.getExpressionNodeDeterminer(),
            ComparatorSingleton.INSTANCE.getNodeValueComparator(),
            atomNodePrinter
        );
        listNotationPrinter = new ListNotationPrinter(
            DeterminerSingleton.INSTANCE.getExpressionNodeDeterminer(),
            atomNodePrinter,
            listNotationExpressionNodePrinter
        );
    }
}

File: src/main/java/com/soapdogg/lispInterpreter/singleton/PrinterSingleton.kt
package com.soapdogg.lispInterpreter.singleton

import com.soapdogg.lispInterpreter.printer.*

enum class PrinterSingleton {
    INSTANCE;

    val atomNodePrinter: AtomNodePrinter = AtomNodePrinter()
    val dotNotationExpressionNodePrinter: DotNotationExpressionNodePrinter
    val dotNotationPrinter: DotNotationPrinter
    val listNotationPrinter: ListNotationPrinter

    init {
        dotNotationExpressionNodePrinter = DotNotationExpressionNodePrinter(
            atomNodePrinter
        )
        dotNotationPrinter = DotNotationPrinter(
            atomNodePrinter,
            dotNotationExpressionNodePrinter
        )
        listNotationPrinter = ListNotationPrinter(
            atomNodePrinter,
            ConverterSingleton.INSTANCE.nodeConverter
        )
    }
}